### PR TITLE
Parser tests

### LINKED ISSUE #43: Operator == false is working, but != true is not
Randomly stumbled over this bug while testing:

For example: `someIdentifier == false` returns true, but the inverse operation `someIdentifier != true` returns false.

The problem might be the evaluation of comparison statements in the interpreter.

### PR DESCRIPTION
**Pull request by marcauberer**
- Add tests for the parser
- Add AST serializer
- Slightly changing the parsing structure of `ComLineBlocks`. Instead of following the grammar straight, the parser parses now directly to the ifBlock, which saves boilerplate code and improves the performance.
- Fix #43

### CODE REVIEW

File: compiler/cpp/src/interpreter/Interpreter.cpp [hunk 1]
```diff
@@ -52,7 +52,11 @@ std::string Interpreter::getOutputOfContent(ContentExprAST* content) {
 }
 
 std::string Interpreter::getOutputOfArbitrarySection(ArbitraryExprAST* arbitraryExpr) {
-    return arbitraryExpr->getValue();
+    std::string arbitrary = arbitraryExpr->getValue();
+    // Cut of first char, if it is a line break
+    if (arbitrary.rfind('\n', 0) == 0)
+        arbitrary.erase(0, 1);
+    return arbitrary;
```
Comment (Splines) [inline compiler/cpp/src/interpreter/Interpreter.cpp:59]: what's the purpose of modifying the arbitrary section here?

(also grammar: use `Cut out`)

File: compiler/cpp/src/parser/ast/ComBlockExprAST.cpp [hunk 2]
```diff
@@ -1,7 +1,24 @@
...
+        case COM_LINE_BLOCK_EXPR: {
+            auto* comLineBlock = static_cast<ComLineBlockExprAST*>(this);
+            return comLineBlock->serialize();
+        }
+        case COM_BLOCK_BLOCK_EXPR: {
+            auto* comBlockBlock = static_cast<ComBlockBlockExprAST*>(this);
+            return comBlockBlock->serialize();
+        }
+        default:
+            throw std::runtime_error("Unknown com block type");
```
Comment (Splines) [inline compiler/cpp/src/parser/ast/ComBlockExprAST.cpp:22]: Shouldn't there be no `ComBlock ExprAST` at all if the type is not correct? But still, good practice to throw an error here.

File: compiler/cpp/test/parser/ParserTest.cpp [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+const ParserParams PARSER_TEST_PARAMETERS[] = {
+        {
+            "has-block"
+        },
+        {
+            "comp-block"
+        },
+        {
+            "contains-block"
+        }
```
Comment (Splines) [inline compiler/cpp/test/parser/ParserTest.cpp:24]: what about ifBlock, stmtList, index, key etc.?

File: compiler/cpp/test/parser/ParserTest.cpp [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+};
+
+class ParserTests : public ::testing::TestWithParam<ParserParams> {};
+
+TEST_P(ParserTests, TestParserWithExpectToken) {
+    ParserParams param = GetParam();
```
Comment (Splines) [inline compiler/cpp/test/parser/ParserTest.cpp:30]: lower-case method names

File: compiler/cpp/test/parser/ParserTest.cpp [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+            "comp-block"
+        },
+        {
+            "contains-block"
+        }
+};
+
+class ParserTests : public ::testing::TestWithParam<ParserParams> {};
+
+TEST_P(ParserTests, TestParserWithExpectToken) {
```
Comment (Splines) [inline compiler/cpp/test/parser/ParserTest.cpp:29]: Nice, these parameterized tests are great 👌 I get why we need the `serialize()` methods now ;)

**Review by Splines — CHANGES_REQUESTED** [review]

File: compiler/cpp/src/interpreter/Interpreter.cpp (hunk 1 above)
Comment (marcauberer) [reply compiler/cpp/src/interpreter/Interpreter.cpp:59]: The problem was that we had unwanted line breaks in the compile output. This line breaks were caused by arbitrary sections that only contain `\n`. This code snippet fixes that issue.

File: compiler/cpp/test/parser/ParserTest.cpp (hunk 3 above)
Comment (marcauberer) [reply compiler/cpp/test/parser/ParserTest.cpp:24]: Those structures are already included in the test files, but I agree that the index token for example is missing in the test files. Will add the index to a test file.

File: compiler/cpp/test/parser/ParserTest.cpp (hunk 4 above)
Comment (marcauberer) [reply compiler/cpp/test/parser/ParserTest.cpp:30]: That is a standardized method name for Google Test, which can not be changed.

**Comment by marcauberer** [comment]
@Splines everything should be fixed now. Can you approve?

File: compiler/cpp/src/interpreter/Interpreter.cpp (hunk 1 above)
Comment (Splines) [reply compiler/cpp/src/interpreter/Interpreter.cpp]: ah ok 👌

and yeah, `cut off` works as well ;)

**Review by Splines — APPROVED** [review]
nice 👌

**Pull request merged**